crates/flashblocks-rpc/src/state.rs

```diff
@@ -0,0 +1,519 @@
+use crate::metrics::Metrics;
+use crate::pending::PendingBlock;
+use crate::rpc::FlashblocksAPI;
+use crate::subscription::Flashblock;
+use alloy_primitives::{Address, TxHash, U256};
+use arc_swap::ArcSwapOption;
+use op_alloy_network::Optimism;
+use reth_optimism_chainspec::OpChainSpec;
+use reth_rpc_convert::RpcTransaction;
+use reth_rpc_eth_api::{RpcBlock, RpcReceipt};
+use std::sync::Arc;
+use std::time::Instant;
+use tokio::sync::broadcast;
+use tokio::sync::broadcast::Sender;
+use tracing::{debug, info, warn};
+
+// Buffer 4s of Flashblocks
+const BUFFER_SIZE: usize = 20;
+
+#[derive(Debug, Clone)]
+pub struct FlashblocksState {
+    pending_block: Arc<ArcSwapOption<PendingBlock>>,
+
+    flashblock_sender: Sender<Flashblock>,
+    metrics: Metrics,
+    chain_spec: Arc<OpChainSpec>,
+}
+
+impl FlashblocksState {
+    pub fn new(chain_spec: Arc<OpChainSpec>) -> Self {
+        Self {
+            pending_block: Arc::new(ArcSwapOption::new(None)),
+            chain_spec,
+            flashblock_sender: broadcast::channel(BUFFER_SIZE).0,
+            metrics: Metrics::default(),
+        }
+    }
+
+    fn is_next_flashblock(
+        &self,
+        pending_block: &Arc<PendingBlock>,
+        flashblock: &Flashblock,
+    ) -> bool {
+        flashblock.metadata.block_number == pending_block.block_number
+            && flashblock.index == pending_block.flashblock_idx + 1
+    }
+
+    fn update_block(&self, block: eyre::Result<PendingBlock>) {
+        let start_time = Instant::now();
+        match block {
+            Ok(block) => {
+                self.pending_block.swap(Some(Arc::new(block)));
+                self.metrics
```

Comment:
im slightly confused about this. its just tracking how much time it takes to do an arc swap?

Reply:
This is a bug, will update -- thanks!

Reply:
This is a bug, will update -- thanks!